### ISSUE TITLE
Circular dependency detection in parser

### ISSUE THREAD

**Issue by jmjwozniak** [swift-lang/swift-t#4]
_From @GoogleCodeExporter on April 22, 2015 18:50_

```
Detect circular dependencies leading to deadlock.

E.g. 
a = b;
b = c;
c = a;

or with conditionals

c = a + 2;
if (...) {
   a = b;
} else {
   a = c;
}

The basic strategy to do this would be to construct a dataflow graph and then 
check for cycles or dead-ends.  E.g. in a procedure there will be some set of 
variables A which start off with an initial value (including arguments, and 
variables which are initialised to a constant value).  If we follow all 
"depends-on" links from a variable v back through the graph, all paths should 
terminate at a variable in the set A.  If there is a cycle, or if the path 
terminates at an uninitialised variable, we have a deadlock.  There are a 
couple of complications I see:
* conditional statements - the value of the variable could come from one branch 
or another, so need to have some kind of OR operator (like the Phi operator 
from SSA) to represent this
* loops - I'm not sure what the implications of these are yet
* assigning to array members.  Sometimes we can work out exactly which array 
index is assigned, but in general we don't know, so we will have to have 
special treatment in the dataflow graph for the unknown array indices
```

Original issue reported on code.google.com by `tim.g.ar...@gmail.com` on 8 Feb 2012 at 3:38
* Blocked on: #78

_Copied from original issue: jmjwozniak/exm-issues#79_

**Comment by jmjwozniak** [comment]
_From @GoogleCodeExporter on April 22, 2015 18:50_

```
Low priority - I don't think this is that important while we're still in the 
research prototype phase
```

Original comment by `tim.g.ar...@gmail.com` on 29 Mar 2012 at 1:36
* Added labels: **Priority-Low**
